src/v/cluster/scheduling/leader_balancer_constraints.h

```diff
@@ -98,6 +101,122 @@ class muted_index final : public index {
+
+        group_info(
+          const raft::group_id& gid,
+          const model::broker_shard& bs,
+          std::vector<model::broker_shard> r)
+          : group_id(gid)
+          , leader(bs)
+          , replicas(std::move(r)) {}
+    };
+
+    using topic_id_t = model::revision_id::type;
+
+    template<typename ValueType>
+    using topic_map = absl::flat_hash_map<topic_id_t, ValueType>;
+
+public:
+    even_topic_distributon_constraint(
+      group_id_to_topic_revision_t group_to_topic_rev,
+      shard_index si,
+      const muted_index& mi);
+
+    even_topic_distributon_constraint(
+      even_topic_distributon_constraint&&) noexcept = default;
+    even_topic_distributon_constraint&
+    operator=(even_topic_distributon_constraint&&) noexcept = default;
+
+    even_topic_distributon_constraint(const even_topic_distributon_constraint&)
+      = delete;
+    even_topic_distributon_constraint&
+    operator=(const even_topic_distributon_constraint&)
+      = delete;
+
+    ~even_topic_distributon_constraint() override = default;
+
+    double error() const { return _error; }
+
+    void update_index(const reassignment& r) override;
+
+    /*
+     * Uses a greedy algorithm to generate a recommended reassignment
+     * that will reduce error for this constraint.
+     *
+     * The gist is that it searches topics in order of how skewed they are.
+     * Then for each topic it finds a partiton on the node with the most
+     * leadership and tries to move it to one of its replica nodes. If this
+     * reassignment reduces error its returned.
+     *
+     * If no reassignment can reduce error for this constraint std::nullopt is
+     * returned.
+     */
+    std::optional<reassignment> recommended_reassignment() override;
+
+private:
+    shard_index _si;
+    std::reference_wrapper<const muted_index> _mi;
+    group_id_to_topic_revision_t _group_to_topic_rev;
+    double _error{0};
+
+    topic_map<absl::flat_hash_map<model::node_id, std::vector<group_info>>>
+      _topic_node_index;
+    topic_map<size_t> _topic_partition_index;
+    topic_map<absl::flat_hash_set<model::node_id>> _topic_replica_index;
+    topic_map<double> _topic_skew;
+    topic_map<double> _topic_opt_leaders;
+
+    const shard_index& si() const { return _si; }
+    const muted_index& mi() const { return _mi.get(); }
+    const group_id_to_topic_revision_t& group_to_topic_id() const {
+        return _group_to_topic_rev;
+    }
+
+    void calc_topic_skew();
+    void rebuild_indexes();
+
+    /*
+     * Compute new error for the given reassignment.
+     */
+    double adjusted_error(
+      double current_error,
+      const topic_id_t& topic_id,
+      const model::broker_shard& from,
+      const model::broker_shard& to) const;
+
+    void calculate_error() {
+        _error = std::accumulate(
+          _topic_skew.cbegin(),
+          _topic_skew.cend(),
+          double{0},
+          [](auto acc, const auto& t_s) { return acc + t_s.second; });
```

Comment:
Instead of `[](auto acc, const auto& t_s) { return acc + t_s.second; }` ? I'm not sure if it's possible to use since I need to access the second member of the `t_s` tuple. I could try `std::views::values` to map the unordered_map to its values first though then use `std::plus` though.

Reply:
> the second member of the t_s tuple

oh yes indeed I missed that, sorry!